File: frontend/src/api/maintenance.ts
const BASE = '/api/maintenance';

export interface DegradationRequest {
  edge_id: string;
  equipment_id: string;
  equipment_type: string;
  design_cop?: number;
  cop_window: number[];
  approach_temp_avg?: number;
  vibration_window?: number[];
}

export interface DegradationResult {
  severity: 'healthy' | 'degrading' | 'critical';
  cop_degradation_pct: number;
  cusum_triggered: boolean;
  recommended_action: string;
}

export interface PredictRequest {
  cop_current: number;
  vibration_rms: number;
  approach_temp: number;
}

export interface PredictResult {
  failure_probability: number;
}

export async function evaluateDegradation(body: DegradationRequest): Promise<DegradationResult> {
  const r = await fetch(`${BASE}/evaluate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error('Failed to evaluate degradation');
  return r.json();
}

export async function predictFailure(body: PredictRequest): Promise<PredictResult> {
  const r = await fetch(`${BASE}/predict`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error('Failed to predict failure');
  return r.json();
}

File: services/agent/agent_service/predictive_maintenance/api/maintenance.py
from fastapi import APIRouter
from pydantic import BaseModel

from ..degradation_tracker import DegradationTracker
from ..failure_predictor import FailurePredictor, build_training_data

router = APIRouter()

# Lazy-init predictor, train on first use
_predictor: FailurePredictor | None = None


def _get_predictor() -> FailurePredictor:
    global _predictor
    if _predictor is None:
        _predictor = FailurePredictor()
        X, y = build_training_data()
        _predictor.train(X, y)
    return _predictor


class DegradationRequest(BaseModel):
    edge_id: str
    equipment_id: str
    equipment_type: str
    design_cop: float = 5.5
    cop_window: list[float]
    approach_temp_avg: float = 0.0
    vibration_window: list[float] = []


class PredictRequest(BaseModel):
    cop_current: float
    vibration_rms: float
    approach_temp: float


@router.post("/evaluate")
async def evaluate_degradation(body: DegradationRequest):
    tracker = DegradationTracker(body.equipment_id, body.equipment_type)
    result = tracker.evaluate(
        design_cop=body.design_cop,
        cop_window=body.cop_window,
        approach_temp_avg=body.approach_temp_avg,
        vibration_window=body.vibration_window,
    )
    result["edge_id"] = body.edge_id
    return result


@router.post("/predict")
async def predict_failure(body: PredictRequest):
    p = _get_predictor()
    proba = p.predict_proba([body.cop_current, body.vibration_rms, body.approach_temp])
    return {
        "failure_probability": round(proba, 4),
        "features": {
            "cop_current": body.cop_current,
            "vibration_rms": body.vibration_rms,
            "approach_temp": body.approach_temp,
        },
    }
